Toggle block comments

gabydd opened the pull request:
this is an initial implementation of toggling block comments which fixes #1505 it still needs
- [ ] decision for keybindings, for now I put block comments under space-c-b and line comments to space-c-c
- [x] if we should have a single command for toggle comment that does either block or line toggle
- [x] add more block comment tokens to language.toml
- [ ] ~can we get comment tokens from treesitter~
- [x] right now toggle a block doesn't extend the selection, should it do that? extends for non linewise block comments
- [x] tests
- [x] comments
- [x] add to docs when keybindings are finalized

any other feedback appreciated, this is still a draft and nothing is set in stone

some implementation notes:
- the command normalizes the commenting across the selections so if one selection is uncommented all selections get commented, this can be changed if needed
- comments surround the content of each selection ignoring whitespace when commenting and leave any whitespace surrounding block comment tokens when uncommenting
- ~~i tried to do the same thing as `find_first_non_whitespace_character` in `find_last_non_whitespace_character` but apparently `Chars` doesn't implement `DoubleEndedIterator` so I just iterate over a reversed range of chars len instead. I don't think this should have performance implications on really long selections, but someone else might know better.~~ pascal figured this out

[video of it in action](https://asciinema.org/a/kkAUgOnELliqhspIUl4Umiy9q)

gabydd commented:
I think I can make it so that there is one command that can toggle block and line comments,, then seperate commands for just block and line toggling, my proposed algorithm is:

if only one of single line token or block tokens exist, toggle that, else do the following:
- check if all selections are line commented
	- if they are then line toggle
- check if all selections are block toggled
	- if they are then line toggle
- check if all selections are single line
	- if they are then line toggle
	- should this check for block comments?
- else toggle block
	- could potentially also check and skip over already line commented code, but then block toggling would also do line toggling

gabydd commented:
I implemented the above, and then separated line comment toggling into it's own command that is under `space+c+l`

there is definitely ways to make block commenting smarter, for example comment a block if selection is surrounded by only whitespace and then comment token or uncomment block that is surrounded by other characters then whitespace in a selection(though I think this could actually extend the comment block maybe). This could be saved for future pr if this functionality is wanted.

pascalkuthe commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +94,168 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+    let (block_commented, _) =
+        find_block_comments(comment_tokens.0, comment_tokens.1, text, selection);
+
+    // line comment when only line comment token available
+    if matches!(token, Some(_)) && !matches!(tokens, Some(_)) {
+        return true;
+        // block comment when only block comment tokens available
+    }
+    if !matches!(token, Some(_)) && matches!(tokens, Some(_)) {
+        return false;
+    }
```
small nit: maybe move this in front of the comment search so that languages with just one type of comment avoid searching for the default comments. As  discussed in the matrix chat it would be nice to specical case languages with just block comments here, so returning an enum would be good. Also I think you can make this clearer as a  match:
``` rust 
let (token, tokens) = match (token, tokens){
 (Some(token), Some(tokens)) => (token, tokens),
 (None, None) => ("//", ("/*","*/")),
 (Some(_), None) => return CommentType::Line,
 (None, Some(_)) => return CommentType::BlockAsLineFallback,
};
```

gabydd replied:
thanks for this review, I implemented it with some changes. The enum and match was a good call here. Turned out ``BlockAsLineFallback`` should only be returned if the selections are only one line long so I changed it to check for that(almost left this for tomorrow but I had an epiphany). I also didn't have the enum variants assigned a value because it wasn't needed and I didn't want to deal with lifetimes. Thank you again for this review it was very helpful.

pascalkuthe reviewed (changes requested):
Thanks for working on this! Comment as discussed in matrix

gabydd commented:
This PR should now be ready for review, some things to note:
- selection isn't extended to encompass the comment, wasn't sure how or if I should do this
- keybindings should be chosen, I know some people wanted to switch to ``#``, I don't have strong opinions about this both are great to me
- this is pretty much first rust stuff I have done, so if anything seems wrong please don't assume it's not you probably know better then me.
- this is breaking change to how comment toggling works
Also want to thank everyone for all the work they have put into helix, using and contributing to it has been a joy
(the doc failing seems to be based on a failed network call, shouldn't effect this pr)

pascalkuthe commented:
Thanks for working on this! I think block comments are useful to have in particular for languages that do not have line comments.
I tried this out a while ago with markdown and it worked well.

I had this in my daily driver for a while and have some thoughts. I think automatically switching between line comments and block commenys based upon whether more then 1 char is selectes is nice to have but I found myself frustrated by this as a default. 

For languages that support them I almost always want line comments because of the more granular control (comment ten lines, later uncomment only two of them or move some part of the comment elsewhere without having to add new commentnstsrt/end) so I would prefer to keep that as is. 
However for languages that do not have line comments by default it should fall back to adding a block comment to every selected lime (same for `<space>cl`).

 I think the automatic option is still useful but perhaphs as a secondary command (maybe mapped to `C-C` by default and people who prefer it the other way around can swap those two?).

If other people (@archseer whatdo you think?) prefer the default as in this PR I think its fine too as the  I keybindings can be swappes in my config. I think the suggesting about using block comments as a line comment fallback should definitely be implement

gabydd commented:
Thanks for your input @pascalkuthe it's very helpful, and unless someone else objects I think I will implement it as the default. This seems simpler and more intuitive to the user, I don't have time today to implement this but it was in the back of my mind after I marked it for review so it shouldn't take me too long.

gabydd commented:
sorry for the wait on this I have been pretty busy, the new relevant changes are more inline with @pascalkuthe suggestions(if you have the time to take a look at the new behaviour that would be appreciated, I am sure you are pretty busy right now so it's totally okay if you don't do an actual review)

the key changes are summarized here:
- toggle comments always does a line comment or uses block as line fallback (this makes it much less of a breaking change)
- it only does a regular block comment when the selection is already block commented

- line comment and block comment now both fallback to each other if there token is not available for that language

whats left at this point is to add this to the docs and look into making some more meaningful tests

gibbz00 commented:
How will this affect the default `"C-c" => toggle_comments` binding?

gabydd commented:
Only effects if there is no line ``comment-token`` for a language(falls back to using the block comment tokens to comment each line instead)or if the whole selection is block commented other then that ``toggle_comments`` stays the same

pascalkuthe commented:
@gabydd Is this ready for review?

pascalkuthe commented:
ok thanks for letting me know. I will think about that during review

gabydd commented:
just fixed the conflict and rebased on master, I realize there is probably some new languages that don't have block comment tokens so will add those

gabydd commented:
there is now documentation as well, I would add block-comments to the tutor but then the user would have to change the language to one that only supports block comments and that seems a little bit confusing for the tutorial

pascalkuthe commented on book/src/keymap.md:
```diff
@@ -289,6 +290,7 @@ This layer is a kludge of mappings, mostly pickers.
 | `h`     | Select symbol references (**LSP**)                                      | `select_references_to_symbol_under_cursor` |
 | `'`     | Open last fuzzy picker                                                  | `last_picker`                              |
 | `w`     | Enter [window mode](#window-mode)                                       | N/A                                        |
+| `w`     | Enter [comment mode](#comment-mode                                      | N/A                                        |
```
```suggestion
| `c`     | Enter [comment mode](#comment-mode                                      | N/A                                        |
```

pascalkuthe commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +98,226 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+        ) {
+            let len = selection_slice.len_chars();
+
+            // selection can be shorter than open_pos + open len
+            let open_fragment = Cow::from(
+                selection_slice.slice(open_pos..std::cmp::min(open_pos + open.len(), len)),
+            );
+            let close_fragment = Cow::from(
+                selection_slice
+                    .slice(close_pos - std::cmp::min(close.len() - 1, close_pos)..close_pos + 1),
+            );
```
no need to create a `Cow` you can compare `RopeSlice` to `str` directly. Also instead of bounding the size, you can simply check upfront that slice is large enough.

You also mix up byte and char indices here. Ropey and helix use char indices by default while strings use byte indices. I this case you should probably just calculate the char len of `open` and `close` at the start of the function and use char indices everywhere. This won't make a difference for ASCII comments but we should correctly handle non-ascii tokens even if they are rare in practice.

gabydd replied:
ah yes this pr was pretty much the first thing I wrote in rust so I just blindly copied this from the line commenting and modified it work for block commenting, looks like there should be a follow up pr to improve the line commenting implementation to take this into account as well.

pascalkuthe replied:
That would be nice, also split selection at whitespace could be improved

pascalkuthe commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +98,226 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+                    .slice(close_pos - std::cmp::min(close.len() - 1, close_pos)..close_pos + 1),
+            );
+
+            // margin of one if there is a space between the comment token and other characters
+            let open_margin = match selection_slice.get_char(open_pos + open.len()) {
+                Some(c) if c == ' ' => 1,
+                _ => 0,
+            };
+
+            let close_margin = if open_pos + open.len()
+                != close_pos - std::cmp::min(close.len(), close_pos)
+            {
+                match selection_slice.get_char(close_pos - std::cmp::min(close.len(), close_pos)) {
+                    Some(c) if c == ' ' => 1,
+                    _ => 0,
+                }
+            } else {
+                0
+            };
```
this code is hard to read. I would simply do this after you have determined that there is indeed a block comment. Then you can calculate the inner slice and use `get_char(0)` and `get_char(slice.len().satusating_sub(1))`. 

you can use `map_or` to make this more readable, also store the margins as a bools. You can then use `as usize` when generating the actual transaction

pascalkuthe commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +98,226 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+    close: &str,
+    text: RopeSlice,
+    selection: &Selection,
+) -> (bool, Vec<CommentChange>) {
```
this could return an iterator I think which would be more efficient/idiomatic

gabydd replied:
I hope I addressed everything else but sadly I don't think we can use an iterator here because we need to check all the selections before we know if we are commenting or not. See how we clear the clear the vec if we find an uncommented selection so that we can normalize all the selections to all be commented.

pascalkuthe replied:
Ah yeah ofcourse I didn't think j about that my bad

pascalkuthe reviewed (changes requested):
I haven't finished yet but I already have some technical comments

gabydd commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +98,226 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
     Transaction::change(doc, changes.into_iter())
 }
 
+fn find_last_non_whitespace_char(text: RopeSlice) -> Option<usize> {
+    (0..text.len_chars()).rev().find(|&i| {
+        text.get_char(i)
+            .map(|ch| !ch.is_whitespace())
+            .unwrap_or(false)
+    })
+}
```
Yeah I tried to do it that way but for some reason using the reversed iterator I was not able to get it to work, maybe I am doing something wrong?

gabydd replied:
I think I was trying something like: ``text.chars().reversed().position(|ch| !ch.is_whitespace())`` which was always returning ``None`` and I had subtracting from the amount of chars as well.

pascalkuthe replied:
Ah right ropey iterator are cursos not double ended like std so you need to do `chars_as(slice.len_chars())` instead of `chars`

gabydd replied:
that allowed me to get it to works thanks!

gabydd commented:
Thank you for the review Pascal ❤️, especially the spliting lines part I thought that might be problematic, I'll try and get to this tonight

gabydd commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +98,226 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+        }),
+        &REGEX,
+    )
+}
```
I believe I was able to implement this correctly, there is no lines iterator from a range, so I just took the a ``RopeSlice`` of the line range and iterated of those lines which worked

pascalkuthe replied:
yeah that's what I meant :+1:

pascalkuthe commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +96,212 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+            let after_open = open_pos + open_len;
+            let before_close = close_pos.saturating_sub(close_len);
+
+            let line_commented = if len >= after_open {
```
```suggestion
            let line_commented = if len >= open_Len + close_len {
```

I think that should ge the right length check?

gabydd replied:
Hmm yeah the current check works for all possibilities but the technically correct check I think would be: ``close_pos - open_pos >= open_len + close_len``

pascalkuthe replied:
Ah I thaught that `len = close_pos -open_pos` but it I counts the trimmed white spaces still. Then that check makes more sense but I would prefer the variant I posted (and change the definition of lrn). It's more obvious what it does I think: Check if the slice is large enough to contain a block comment. The current condition could also panic for non-symmetric block comments where the open brace is shorter. HTML jas non-symetric block comments (altough there the open brave is longer) so it's not completely infeasiabke that we would auoprt such comments in the future (or that some user configures the editor that way)

gabydd replied:
Change made👍

gabydd commented:
I think I will probably have to clean up the history of this pr because it is getting quite messy, I am thinking one commit for the actual commenting and another for all the ``languages.toml`` changes

pascalkuthe commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +96,212 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+    }
+    let split_lines = split_lines_of_selection(text, selection);
+    let (line_commented, block_commented) = match (token, tokens) {
+        (Some(_), Some(tokens)) => (
+            find_block_comments(tokens.0, tokens.1, text, &split_lines).0,
+            find_block_comments(tokens.0, tokens.1, text, selection).0,
+        ),
+        (None, None) => (
+            find_block_comments("/*", "*/", text, &split_lines).0,
+            find_block_comments("/*", "*/", text, selection).0,
+        ),
+        (Some(_), None) => return CommentType::Line,
+        (None, Some(tokens)) => (
+            find_block_comments(tokens.0, tokens.1, text, &split_lines).0,
+            find_block_comments(tokens.0, tokens.1, text, selection).0,
```
you are essentially looking for block comments here and immedietly throwing away the result. I think there should we a second function for creating a transaction for block comment. Then you can have the following commands:
* toggle block comments (calls find_block_comment and create_block_comments_transaction)
* toggle comments: basically this function but instead of returning an enum immediately call `toggle_block_comments_as_line_fallback`, `toggle_line_comments` or `create_block_commen_transaction` (reusing the resunts of `find'block_comments` from inside this function)

gabydd replied:
Ah that makes sense, I was a bit iffy about doing this as well and that works much better

gabydd replied:
this is now implemented and I was also able to simplify the detection for the toggle_comments command, i then moved it into ``commands.rs`` I can move it back if you prefer though

pascalkuthe replied:
i think it looks great lke this. Great split: core logging in helix-core glue code in the command implementation

pascalkuthe reviewed (approved):
From a technical point of view this LGTM now thanks!

>  or do we think that can be a separate option?

I see your point I think having the continue block comment token in the middle would indeed be nice. So `["/*" ,"  *" ,"/*"]`. On the other hand having it as a separate config works too and would avoid having to make a decision now/update all configs. I am also wondering if we somehow  want to support multiple comment tokens per language (so we can support doc comments for example, there are block doc comments in rust that would have a different continuation token). 

I am not quite sure about the config. If we can't settle on something at all now I would lean towards merging as is and then adding a separate config option for continuation later

pascalkuthe commented:
I discussed with @the-mikedavis how the config should look like to be future proof and we cameup with the following (example config for rust):

``` toml
comment-tokens = ["//", "///", "//!"]
block-comment-tokens = [
  { start = "/*", end = "*/" },
  { start = "/**", end = "*/" },
  { start = "/*!", end = "*/" },
]
```

So the core idea is that:
* we want to be able to uncomment multiple different kinds of comments so we want a list of token (I think for commenting we can always just use the first token for now. This is especially important for continuing comments correctly (which shouldn't be too hard to implement with this config)
* we want to be able to add somekin of config option to block comments in the future for how to continue comments. We don't really want to settle on that right now and using a dict makes it possible to add that token in the future without a breaking change. It also read a little nicer IMO

To stay backwards compatible (and make the config for block comments a little less cluttered since most languages just have a single block comment type) we want to be able to also just specify a string/a dict derctly instead of a single element list:

``` toml
comment-tokens =  "//"
block-comment-tokens = { start = "/*", end = "*/" }
```

This should also mean that you can very quickly update the languages.toml with multi cursors instead of having to go through it again by hand (altough if would be nice if you would adopt my rust example).

Implementing this kind of config option can be done with an untagged enum in serde we do this for quite a few config options so you can use that as reference

pascalkuthe commented:
ah sorry one thing I missed:

>  right now toggle a block doesn't extend the selection, should it do that?

Imo yes it should but not for the line comment fallback (since c-c doesn't change the selection either) but for the `<space>cb` binding it should

gabydd commented:
Both of that sound good to me. Just a quick clarification, this pr should also update the comment-tokens config aswell, right?

pascalkuthe commented:
> Both of that sound good to me. Just a quick clarification, this pr should also update the comment-tokens config aswell, right?

yeah ideally we would do it in this PR since we don't want a breaking change later

gabydd commented:
ok I have now implemented the new config, the way it works it will search each range to see if it is commented, if all ranges are commented it will uncomment them all, if atleast one is not commented that range will be block commented with either the first token in the array or the default tokens, next I need to resolve the conflicts with master, and try and extend the selection. the history should also probably be squashed at some point.

gabydd commented:
I found a way to do the extending when block commenting not sure if it is the best way just the easiest way I could figure out: [`2db9dea` (#4718)](https://github.com/helix-editor/helix/pull/4718/commits/2db9deaac8829af823161f24b05f4a550699c748), maybe if the selection should be extended should be passed as an argument to ``create_block_comment_transaction`` so the ranges are never returned?

gabydd commented:
ah one bug in the implementation of extending is it will panic when there is an only whitespace selection, I will have to fix that sometime in the morning though

pascalkuthe commented on book/src/languages.md:
```diff
@@ -61,7 +61,8 @@ These configuration keys are available:
 | `auto-format`         | Whether to autoformat this language when saving               |
 | `diagnostic-severity` | Minimal severity of diagnostic for it to be displayed. (Allowed values: `Error`, `Warning`, `Info`, `Hint`) |
-| `comment-token`       | The token to use as a comment-token                           |
+| `comment-tokens`      | The tokens to use as a comment token, either a single token `"//"` or an array `["//", "///", "//!"]` (the first token will be used for commenting) |
```
While it doesn't sound as niceas the plural version  I would prefer we keep `comment-token` to avoid a breaking change.

pascalkuthe replied:
Ah actually you can keep the change and use `serde(alias = 'comment-token')` yo keep legacy configs working

pascalkuthe commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +97,206 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+
+            for BlockCommentToken { start, end } in &tokens {
+                let len = (end_pos + 1) - start_pos;
+                let start_len = start.chars().count();
```
Tiny not, move this outside if the loop.

The conouker likely dies that anyway but I personally thjbj it makes the code easier to read if inky the stuff that actually needs to be calculated in a loop is inside the loop

gabydd replied:
moved len outside the loop :+1:

pascalkuthe commented on helix-core/src/comment.rs:
```diff
@@ -94,6 +97,206 @@ pub fn toggle_line_comments(doc: &Rope, selection: &Selection, token: Option<&st
+        if let (Some(start_pos), Some(end_pos)) = (
+            find_first_non_whitespace_char(selection_slice),
+            find_last_non_whitespace_char(selection_slice),
+        ) {
```
Regarding your crash: You need to bail our here if `start_pos >= end_pos`

gabydd replied:
the crash was actually because I wasn't including the only whitespace ranges in the ``comment_changes`` vec so they weren't being added to the new selection that was constructed when commenting so there could be less ranges then the primary index of the selection. I changed the ``CommentChange`` struct to be an enum so all whitespace ranges could be included to be manually mapped

pascalkuthe reviewed (commented):
Thus looks very promising, just a couple smaller comnents but this LGTM for the most part 👍

the-mikedavis commented on book/src/keymap.md:
```diff
@@ -299,6 +301,16 @@ This layer is a kludge of mappings, mostly pickers.
+
+| Key     | Description                          | Command               |
+| -----   | -----------                          | -------               |
+| `c`     | "Comment/uncomment selections"       | toggle_comments       |
+| `l`     | "Line comment/uncomment selections"  | toggle_line_comments  |
+| `b`     | "Block comment/uncomment selections" | toggle_block_comments |
```
```suggestion
| `c`     | Comment/uncomment selections       | `toggle_comments`       |
| `l`     | Line comment/uncomment selections  | `toggle_line_comments`  |
| `b`     | Block comment/uncomment selections | `toggle_block_comments` |
```

for consistent formatting with the other keymap docs

the-mikedavis commented on languages.toml:
To cut down on the size of the diff, what do you think about restoring changes where `comment-token` was just a single token? For example `sway`'s comment-tokens change could be reset

gabydd replied:
Yeah I can do that, that makes sense

the-mikedavis reviewed (commented):
Just a few small comments from an initial read-through. I'll merge this to my driver branch and circle back for a proper review once I give it a try

the-mikedavis commented on helix-core/src/syntax.rs:
```diff
@@ -231,6 +233,35 @@ impl<'de> Deserialize<'de> for FileType {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+#[serde(untagged, rename_all = "kebab-case", deny_unknown_fields)]
+pub enum CommentTokens {
+    Mutliple(Vec<String>),
+    Single(String),
+}
```
These untagged enums can feel a little clunky to work with. If we do a little more work during deserialization we can treat this type like a `Vec<String>` which is a nicer API when you want to iterate over the tokens or grab the first. We can also assert as deserialization-time that the Vec can't be empty:

```rust
#[derive(Debug, Serialize)]
pub struct CommentTokens(Vec<String>);

impl<'de> Deserialize<'de> for CommentTokens {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
        struct Visitor;

        impl<'de> serde::de::Visitor<'de> for Visitor {
            type Value = Vec<String>;

            fn expecting(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
                write!(
                    formatter,
                    "a sequence of comment tokens or a single comment token"
                )
            }

            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
            where
                E: serde::de::Error,
            {
                Ok(vec![v.to_string()])
            }

            fn visit_seq<A>(self, mut seq: A) -> Result<Self::Value, A::Error>
            where
                A: serde::de::SeqAccess<'de>,
            {
                let mut tokens = Vec::new();

                while let Some(token) = seq.next_element()? {
                    tokens.push(token);
                }

                if tokens.is_empty() {
                    Err(serde::de::Error::custom(
                        "`comment-tokens` must not be empty",
                    ))
                } else {
                    Ok(tokens)
                }
            }
        }

        deserializer.deserialize_any(Visitor).map(CommentTokens)
    }
}

impl ops::Deref for CommentTokens {
    type Target = Vec<String>;

    fn deref(&self) -> &Self::Target {
        &self.0
    }
}
```

the-mikedavis replied:
`BlockCommentTokens` can use this same pattern but instead of a `visit_str` it will want a `visit_map` that delegates to the MapAccess deserializer (which will automagically deserialize the BlockCommentToken struct)

```rust
fn visit_map<A>(self, map: A) -> Result<Self::Value, A::Error>
where
    A: MapAccess<'de>,
{
    let deserializer = serde::de::value::MapAccessDeserializer::new(map);
    Deserialize::deserialize(deserializer)
}
```

the-mikedavis replied:
Ah also looks like there's a typo, mutliple => multiple :P

pascalkuthe replied:
IMO its cleaner to implement the deserializer with an internal untagged enum:

``` rust 
impl<'de> Deserialize<'de> for CommentTokens {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: serde::Deserializer<'de>,
    {
		#[derive(Deserialize)]
		#[serde(untagged)]
	    enum CommentTokens {
		    Mutliple(Vec<String>),
		    Single(String),
		}
		match CommentTokens::deserialize(deserializer){
			Single(val) => Self(vec![val]),
			Multile(vals) => Self(vals),
		}
	}
}
```

Since we don't really need a cust type using `#[serde(from)]` is even cleaner:

``` rust
#[derive(Deserialize)]
#[serde(untagged)]
enum CommentTokens {
    Mutliple(Vec<String>),
    Single(String),
}
impl From<CommentTokens> for Vec<String> {
    fn from(value: CommentTokens) -> Self {
        match value {
            CommentTokens::Mutliple(vals) => vals,
            CommentTokens::Single(val) => vec![val],
        }
    }
}
```

then you can just have a smile vec field

``` rust
#[serde(from = "CommentTokens"))]
comment_tokens: Vec<String>
```

gabydd replied:
turns out ``from`` can only be used for container attributes https://serde.rs/container-attrs.html so I used ``deserialize_with`` instead

pascalkuthe replied:
Ah right I gorogt about that. I actually almost prefer this tough since it makes it cleat that the enun is just a deserlaize helper 👍

the-mikedavis commented on helix-term/src/commands.rs:
```diff
@@ -4533,18 +4535,122 @@ pub fn completion(cx: &mut Context) {
+        let text = doc.slice(..);
+
+        // only have line comment tokens
+        if matches!((token, tokens.clone()), (Some(_), None)) {
```
This can also be used for the other `matches!` in this file

the-mikedavis commented on helix-term/src/commands.rs:
```diff
@@ -4533,18 +4535,122 @@ pub fn completion(cx: &mut Context) {
 // comments
-fn toggle_comments(cx: &mut Context) {
+type CommentTransactionFn = fn(
+    token: Option<&str>,
+    tokens: Option<Vec<BlockCommentToken>>,
```
I think it might be clearer to call these `line_token` and `block_tokens`

the-mikedavis commented on helix-term/src/commands.rs:
```diff
@@ -4533,18 +4535,122 @@ pub fn completion(cx: &mut Context) {
+        let split_lines = comment::split_lines_of_selection(text, selection);
+
+        let (line_commented, line_comment_changes) = comment::find_block_comments(
+            tokens
+                .clone()
+                .unwrap_or_else(|| vec![BlockCommentToken::default()]),
```
Here and in the `comment` module we can remove most of these clones by passing around slices of the Vec, so `&[BlockCommentToken]` rather than `Vec<BlockCommentToken>`. `comment::find_block_comments` will need to switch the line `let mut tokens = tokens` to `let mut tokens = tokens.to_vec()` which will allocate a new vec but we can get away with not allocating the vec otherwise.

Instead of `unwrap_or_else` in this file we will need:

```rust
let default_block_comment_tokens = &[BlockCommentToken::default()];
let block_comment_tokens = tokens.unwrap_or(default_block_comment_tokens);
```

so that the slice lives long enough

the-mikedavis reviewed (commented):
Behavior-wise this is looking good to me. Just a few small nits and a suggestion about vecs vs slices

archseer commented on helix-term/src/keymap/default.rs:
```diff
@@ -274,6 +274,11 @@ pub fn default() -> HashMap<Mode, KeyTrie> {
             "k" => hover,
             "r" => rename_symbol,
             "h" => select_references_to_symbol_under_cursor,
+            "c" => { "Toggle comments"
+                "c" => toggle_comments,
+                "l" => toggle_line_comments,
+                "b" => toggle_block_comments,
+            },
```
How about using `space c` for toggle, `space C` for block, then `space alt-C` for block? That way we make it a simple press for the default usecase, and the other ones can be accessed with modifiers

gabydd replied:
that seems good could you clarify which one you would like to be line comment?

gabydd replied:
I changed it to being alt-c for line comment, no worries if you want it the other way but I imagine more people would use block comment binding then the line comment one.

archseer commented on helix-term/src/commands.rs:
```diff
@@ -4533,18 +4535,118 @@ pub fn completion(cx: &mut Context) {
     exit_select_mode(cx);
 }
 
+fn toggle_comments(cx: &mut Context) {
```
This needs some documentation on how it works. If I'm commenting a fresh set of lines, does it use block or line comments? I'd prefer it to use line comments if it can, because it makes it easier to partially uncomment

archseer replied:
Also, let's say that I'm trying to uncomment two lines with line comments, I could start my selection in the middle of the first line, then end in the middle of the second line and both lines would uncomment (in other words I don't have to do a perfect selection). Is that behaviour retained?

gabydd replied:
ah yes thats a good idea, I'll do that now, but the idea of this is it doesn't change anything at all if there is a line comment, if the whole selection is surrounded in block comment then it will uncomment that region, if the whole selection is line commented the uncomment that, if there is no comments or not all lines are commented with block comments and there is only block comment tokens it will comment each non commented line individually with those tokens, if each line is commented with line comments then it will un-comment all the lines.

gabydd replied:
this is how I just commented it:
```rust
/// commenting behavior:
/// 1. only line comment tokens -> line comment
/// 2. each line block commented -> uncomment all lines
/// 3. whole selection block commented -> uncomment selection
/// 4. all lines not commented and block tokens -> comment uncommented lines
/// 5. no comment tokens and not block commented -> line comment
```

nkitsaini commented:
Just curious if/how this can handle the case where same file type requires different types of comment characters. For example `html` files can require `//` for `<script>`, `/* */` for `<style>` and `<!-- -->` for html. Currently I could not find a way to manually map keybindings to specific comment characters.

pascalkuthe commented:
that's out of scope for this PR, HTML is really weird in that regard, the best way would likely to handle that using injections in the future

the-mikedavis commented:
That's https://github.com/helix-editor/helix/issues/7364

gabydd commented:
Just fixed the merge conflicts I also squashed the pr to one commit so it's easier to rebase when there are conflicts if we want I can make one commit the code changes and the other the languages.toml changes

pascalkuthe commented:
(Also closes #4009 but I can't link the issue because the GH UI is terribble)

gabydd commented:
This only fixes that for block comments, I didn't want to change the functionality of line comments in this pr

gabydd commented:
What we could do is for languages without a block comment token, when using the toggle_block_comments it would only add or remove from the start of the selection as if you had a block comment that was the line comment on the left and an empty string on the right

gabydd commented:
redundant_clone seems to have some false positives: https://github.com/rust-lang/rust-clippy/issues/10577

the-mikedavis commented:
Let's add some `#[allow(..)]`s arround those for now with a reminder to check if we can remove them when the MSRV is bumped. Looks like the issue isn't solved yet so we might have to keep the `allow`s around for a little while

daedroza commented:
Using this since last week and works great, thanks for putting this up!

David-Else commented:
Awesome merge!

> How will this affect the default `"C-c" => toggle_comments` binding?

It still seems to live on like a Zombie refusing to die, can we bury it now?
